src/mds/Locker.cc

```diff
@@ -295,19 +310,30 @@ bool Locker::acquire_locks(MDRequestRef& mdr,
 	  lov.add_xlock(&dn->versionlock, i + 1);
 	}
       }
-      if (lock->get_type() >= CEPH_LOCK_IFIRST && lock->get_type() != CEPH_LOCK_IVERSION) {
-	// inode version lock?
-	CInode *in = static_cast<CInode*>(object);
-	if (!in->is_auth())
-	  continue;
-	if (mdr->is_leader()) {
-	  // leader.  wrlock versionlock so we can pipeline inode updates to journal.
-	  lov.add_wrlock(&in->versionlock, i + 1);
-	} else {
-	  // peer.  exclusively lock the inode version (i.e. block other journal updates).
-	  // this makes rollback safe.
-	  lov.add_xlock(&in->versionlock, i + 1);
-	}
+      if (is_inode_lock(t)) {
+        switch (t) {
+          case CEPH_LOCK_IVERSION:
+          case CEPH_LOCK_IQUIESCE:
+            break;
+          default:
+	    CInode *in = static_cast<CInode*>(object);
+            if (need_quiescelock) {
+              need_quiescelock = false;
+              lov.add_rdlock(&in->quiescelock);
```

Comment:
I don't understand how this will work for a `handle_client_mksnap` which will first try to acquire an xlock on the snaplock on the same directory that's quiesced:

```
  // lock snap
  if (!(mdr->locking_state & MutationImpl::ALL_LOCKED)) {
    MutationImpl::LockOpVec lov;
    lov.add_xlock(&diri->snaplock);
    if (!mds->locker->acquire_locks(mdr, lov))
      return;
```

Reply:
This is precisely why `quiesce_subvolume` does not xlock the quiescelock:

https://github.com/ceph/ceph/pull/54581/files/607ea4fd57e7459d4b2a81e5c4ecba9680f4d799#diff-ea7cb1a6ba9fa08363b14dd00a86bc6b79e01673e93af84ffdfcdbd0d3f26b19R13570

Reply:
hmm. That looks non-uniform. Maybe we should rather consider setting a flag on the mdr, marking requests that shouldn't attempt to lock the quiescelock? There might be other operations that we'd like to allow for a quiesced root. The `need_quiescelock` would be initialized with that mdr field in mind then, instead of being `true` unconditionally.

Reply:
what happens if we attempt to quiesce the same root again?

Reply:
> hmm. That looks non-uniform. Maybe we should rather consider setting a flag on the mdr, marking requests that shouldn't attempt to lock the quiescelock? There might be other operations that we'd like to allow for a quiesced root. The `need_quiescelock` would be initialized with that mdr field in mind then, instead of being `true` unconditionally.

That could be done. It moves the magic to the requests then which I don't know if I like that any better.

Reply:
> what happens if we attempt to quiesce the same root again?

It would block waiting for the previous quiesce to complete.

Reply:
>  It moves the magic to the requests then which I don't know if I like that any better.

There's already magic in the locker that grabs the quiesce lock for reads. It could be reasonably to encapsulate it all there and control it with the flags on the MDR

Reply:
> > It moves the magic to the requests then which I don't know if I like that any better.
> 
> There's already magic in the locker that grabs the quiesce lock for all other operations. It could be reasonable to encapsulate it all there and control it with the flags on the MDR
> 
> > It would block waiting for the previous quiesce to complete.
> 
> I'm not sure I agree with the semantics. Just like adding the same root to multiple quiesce sets is not resulting in a timeout, we should probably allow roots to be quiesced with multiple requests by a shared write lock semantics on the quiesce lock.. We should discuss this further but I'm fine with leaving the fix for another iteration.

The agent should deduplicate requests. It is not trivial to make the locks work the way you propose.

> However, what I was referring to here was that for a root `a` we take all the quiesce locks of `a/...` but not for the `a` itself. That seems odd.

It's strange. We could perhaps change `mksnap` to signal the Locker to not add `quiescelock` instead.

Reply:
> It's strange. We could perhaps change mksnap to signal the Locker to not add quiescelock instead.

Yes! That's exactly what I meant by `control it with the flags on the MDR`. MDR should carry this flag for the locker to know how and whether to take the quiesce lock for this request. The default is to take it for read, the quiesce will take it exclusively, and mksnap - and maybe other operations we are yet to discover - will want to avoid locking it whatsoever

Reply:
I understood what you meant but I think it suffices to keep it simple as a parameter to `Locker::acquire_locks`. No need to carry a flag in the `MutationImpl`. `mksnap` does no path traversal and simply acquires the `snaplock` currently.

Reply:
Backtracking... we cannot easily do this because `lookup ino=<subvolino>` prior to `mksnap` acquires the `quiescelock`. So we don't even get that far. It may be possible to resolve this in a clean way but it'll have to be for a future refactor/cleanup. I'll add a comment.

Reply:
When does the lookup happen? is it part of the `handle_client_mksnap`?

Reply:
The `Client` issues the `lookup` to resolve the inode number prior to sending `mksnap`.

Reply:
hm... and what capabilities does the client request along with the lookup? Do we have to prevent a lookup on a node that's quiesced?

Reply:
I don't have it off hand but you can test it trivially by turning on debugging, quiescing a subvolume, and then trying to access the subvolume again through some subvolume API (like  `getpath`).